public/app/plugins/panel/heatmap/heatmap_data_converter.ts

```diff
@@ -51,6 +56,39 @@ function elasticHistogramToHeatmap(seriesList) {
   return heatmap;
 }
 
+/**
+ * Sort series representing histogram by label value.
+ */
+function sortSeriesByLabel(s1, s2) {
+  let label1, label2;
+
+  try {
+    // fail if not integer. might happen with bad queries
+    label1 = parseHistogramLabel(s1.label);
+    label2 = parseHistogramLabel(s2.label);
+  } catch (err) {
+    console.log(err);
+    return 0;
+  }
+
+  if (label1 > label2) {
+    return 1;
+  }
+
+  if (label1 < label2) {
+    return -1;
+  }
+
+  return 0;
+}
+
+function parseHistogramLabel(label: string): number {
+  if (label === '+Inf') {
```

Comment:
Mapping `+Inf` to Infinity should be done in the Prometheus datasource.

Reply:
That's for datasources that don't return sorted series. Maybe I should explicitly do it only if datasource is not a prometheus or elastic.